### PR TITLE
feat(otel-node): handle enable/disable instrumentations via env vars

### PR DESCRIPTION
**Pull request by david-luna**

### CODE REVIEW

File: packages/opentelemetry-node/lib/instrumentations.js [hunk 1]
```diff
@@ -180,7 +206,39 @@ function getInstrumentations(opts = {}) {
...
+        if (disabledList.length && disabledList.includes(name)) {
+            return;
+        }
+
+        if (enabledList.length && !enabledList.includes(name)) {
+            return;
+        }
+
+        // 2nd check the configuration passed
+        // XXX: discuss what takes precedence if `opts` argument id passed in
```
Comment (david-luna) [inline packages/opentelemetry-node/lib/instrumentations.js:241]: note for reviewer: leave your thoughts here :)

File: packages/opentelemetry-node/test/OTEL_NODE_ENABLE_DISABLE_INSTRUMENTATIONS.test.js [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                getLine(
+                    'Enabling instrumentation \\"@opentelemetry/instrumentation-ioredis\\"'
+                ),
+                'should not enable instrumentations that are not defined in env vars'
+            );
+        },
+    },
+];
+
+test('OTEL_NODE_RESOUCE_DETECTORS', (suite) => {
```
Comment (trentm) [inline packages/opentelemetry-node/test/OTEL_NODE_ENABLE_DISABLE_INSTRUMENTATIONS.test.js:116]:
```suggestion
test('OTEL_NODE_ENABLED_INSTRUMENTATIONS, OTEL_NODE_DISABLED_INSTRUMENTATIONS', (suite) => {
```

File: packages/opentelemetry-node/lib/instrumentations.js (hunk 1 above)
Comment (david-luna) [reply packages/opentelemetry-node/lib/instrumentations.js]: if I got it right form our last meeting I've done the changes 
- env vars take precedence over config
- if env var enables a instrumentation disabled by config it becomes enabled

File: packages/opentelemetry-node/lib/instrumentations.js [hunk 3]
```diff
@@ -131,6 +134,29 @@ const INSTRUMENTATIONS = {
...
+ * list for OTEL environment vars. Example:
+ * https://opentelemetry.io/docs/languages/sdk-configuration/general/#otel_propagators
+ *
+ * If the param is not defined or falsy it returns an empty array. The resulting
+ * array has only nonempty string.
+ *
+ * @param {string | undefined} str
+ * @returns {Array<string>}
+ */
+function parseStringList(str) {
```
[the remark this answers is not in the document]
Comment (david-luna) [reply packages/opentelemetry-node/lib/instrumentations.js]: I ended updating it to something more complex

File: packages/opentelemetry-node/lib/instrumentations.js [hunk 4]
```diff
@@ -179,22 +215,48 @@ const INSTRUMENTATIONS = {
...
+    );
 
     Object.keys(INSTRUMENTATIONS).forEach((name) => {
+        // Skip if the env configuration says so
+        if (enabledFromEnv && !enabledFromEnv.includes(name)) {
+            return;
+        }
+        if (disabledFromEnv && disabledFromEnv.includes(name)) {
+            // Do not skip if it was in the enable list
+            if (!enabledFromEnv) {
```
Comment (trentm) [inline packages/opentelemetry-node/lib/instrumentations.js:232]: I think this is wrong. If an instrumentation name is listed in both OTEL_NODE_ENABLED_INSTRUMENTATIONS and OTEL_NODE_DISABLED_INSTRUMENTATIONS, then it should be *disabled*. I.e. disabling wins. That is what auto-instr-node is doing:

https://github.com/open-telemetry/opentelemetry-js-contrib/blob/60a99c98fd3a5594c7c2234184f06166b375e707/metapackages/auto-instrumentations-node/test/utils.test.ts#L117-L134

**Review by trentm — CHANGES_REQUESTED** [review]
One change to match the auto-instr-node behaviour (DISABLED envvar wins). Otherwise looks good.

**Review by trentm — APPROVED** [review]

**Pull request merged**